File: testdata/fixtures/ocaml/CWE-22b.ml
(* CWE-22: Path Traversal *)
let vuln user =
  (* ruleid: ocaml-path-traversal *)
  let _ = Filename.concat "/tmp" user in
  (* ruleid: ocaml-path-traversal *)
  let ch = open_in user in
  ()
let safe () =
  (* ok: ocaml-path-traversal *)
  let _ = Filename.concat "/tmp" "fixed.txt" in
  (* ok: ocaml-path-traversal *)
  let ch = open_in "/tmp/fixed.txt" in
  ()
let vuln2 user =
  (* ruleid: ocaml-path-traversal *)
  let _ = open_in ("/tmp/" ^ user) in
  (* ruleid: ocaml-path-traversal *)
  let _ = In_channel.input_all user in
  ()
let safe2 () =
  (* ok: ocaml-path-traversal *)
  let _ = Filename.concat "/tmp" "safe2.txt" in
  let _ = open_in "/tmp/safe2.txt" in
  ()

File: testdata/fixtures/elixir/CWE-22b.ex
# CWE-22: Path Traversal via File.read(Path.join)
defmodule Fixture.CWE22b do
  def vuln1(conn) do
    user = conn.params["file"]
    # ruleid: elixir-path-traversal
    File.read(Path.join("/tmp", user))
  end

  def vuln2(conn) do
    path = conn.query_params["path"]
    # ruleid: elixir-path-traversal
    File.read!(Path.join("/data", path))
  end

  def vuln3(conn) do
    name = conn.params["name"]
    p = Path.join("/srv", name)
    # ruleid: elixir-path-traversal
    File.read(p)
  end

  def safe do
    # ok: elixir-path-traversal
    File.read("/tmp/fixed.txt")
  end

  def safe2 do
    # ok: elixir-path-traversal
    File.read(Path.join("/tmp", "fixed.txt"))
  end
end
